libs/langchain/langchain/tools/gmail/get_message.py

```diff
@@ -33,6 +33,27 @@ def _run(
         run_manager: Optional[CallbackManagerForToolRun] = None,
     ) -> Dict:
         """Run the tool."""
+
+        def decode_payload(payload: bytes, charset: Optional[str] = None) -> str:
```

Comment:
Code looks good, but do you have unit tests by any chance? It'd be nice to pull out the common code since it's duplicated across get_message and search.py and isn't tested

Reply:
My apologies. Could you kindly point me to the folder/files in which to write the tests? I saw many under langchain/tests/unit_tests, but I'm not sure where exactly.

Reply:
The best thing to test here is the parser (not the request to gmail). So we can do this in unit-tests, generally it makes sense to mirror the same file structure as where the code is located.

For example:

```
tests/unit_tests/tools/gmail.py 
```

Reply:
Thanks, I've added it (hope it's o.k - noob here)